### DISCUSSION TITLE
/Applications/UTM.app/Contents/MacOS/utmctl start HAOS - Not finding VM error only on startup

### DISCUSSION THREAD

**Discussion by EagleDTW** [utmapp/UTM#6367] [UTM Help (macOS)] [answered]
**Describe the issue**  
This works in shortcuts and in terminal, but not on startup login shortcut shell script:
/Applications/UTM.app/Contents/MacOS/utmctl start HAOS

Says that it can't find the VM location...Is there something better to use such as UUID and how do you find this for the VMs?

**Configuration**  
* UTM Version: 4.5.2
* macOS Version: 14.4.1
* Mac Chip (Intel, M1, ...): M2 Pro

**Crash log**  
N/A

**Debug log**  
N/A

**Upload VM**  
N/A

**Comment by EagleDTW** [comment]
![Screenshot 2024-05-18 at 11 09 30 AM](https://github.com/utmapp/UTM/assets/62778358/55e7a521-304b-4b17-b071-b2dc993d58ef)

**Comment by osy** [comment]
Run `/Applications/UTM.app/Contents/MacOS/utmctl` to see all the options and `/Applications/UTM.app/Contents/MacOS/utmctl list` to get a list of VMs and their UUID.

**Comment by EagleDTW** [comment]
Hi Osy, using either the name of the VM or the UUID does not work and has an error stating that itcannot locate the VM. Please help, is there further settings required in security to enable this? The shortcut works when not used as startup shortcut.

**Comment by EagleDTW** [comment]
@osy Is there any resolution on this...any solutions to have the VM located on startup...like open utm with separate shortcut and then wait 30 seconds before issuing the command? Just an FYI - I tried this and it didn't work! and if running as administrator requires an approval for adding a helper at every startup, is there a way around this?

**Comment by EagleDTW** [comment]
is this an issue for others or only me?

**Reply by EagleDTW** [reply]
@osy still an issue in 4.5.3, is this actively being worked on and am I missing a simple config change that would make it operational?

**Reply by eccgecko** [reply]
Yeah I’m having this issue too, and it’s extremely annoying. Can’t autostart VM’s when machine boots up, which is far from ideal for headless setups.

Weird thing for me is that this used to work fine, and as far as I can tell it’s just stopped working for no apparent reason.

**Comment by EagleDTW** [comment]
Could this be because I have autologin option set for my user in the user section and for some reason it's not passing the permissions for accessing the VM location correctly?

**Reply by EagleDTW** [reply]
@osy is there anyone else having this issue where the VM is not located with an error message on login? I have autologin enabled and the shortcut works just fine normally when the OS is fully up. This renders my usability of UTM as a recovery on power outages and remote reboots moot without this functionality, which was one of the big draws in the first place to buy UTM!

**Reply by ku1ik** [reply]
@EagleDTW please don't ping random people to escalate your issue. Thanks.

**Reply by eccgecko** [reply]
@ku1ik doesn’t seem to me that @EagleDTW did ping “random people”. They were asking someone who has previously replied to this thread whether not anyone else is experiencing this extremely debilitating bug, as it renders UTM useless on headless setups where you want a VM to auto start upon boot (and they aren’t the only one - I am also experiencing this and how I came to find this thread).

Would be nice if UTM actually responded to the substance of @EagleDTW ‘s posts, rather than just chastise them for asking a question of someone who had previously had input into this GitHub issue.

**Reply by ku1ik** [reply]
@eccgecko they did ping me and several others (the post was edited after my comment above), and I never posted anything in this discussion, neither I'm a developer of this project. I don't care about this bug, please stop wasting my time by bringing me into this discussion.

**Reply by eccgecko** [reply]
Ah OK, my mistake. I didn’t see the unedited original, so I’m sorry for that. Reading the thread as a newcomer, it looked like you were saying that @EagleDTW pinging @osy, who previously replied to this thread and changed it from an issue to a discussion, was them pinging random people.

**Comment by eccgecko** [comment]
Just want to make a separate reply to say that I am also experiencing this.

I previously had a macOS shortcut that used the “utm://start?name=…” url option to autostart a headless VM I have on my M1 Mac mini upon macOS startup. This seemingly randomly stopped working.

Like @EagleDTW I have tried using utmctl to get this function back, but have similarly found that it only works when running the command manually in Terminal once the computer is up and running. When the utmctl command is run by Shortcuts upon startup, regardless of whether the VM Name or UUID is used, it does not work and reports that the VM cannot be found. Also, the “utm://start…” url function has apparently stopped working entirely for me, whether or not that is upon startup or when run manually after startup.

**Reply by tomtomtls** [reply]
Same for me on Sonoma + UTM. Now I don't know how to autostart a VM after a reboot (without having to log in).

**Reply by EagleDTW** [reply]
Hi @osy, do we have any current information on this defect? Is there an update that will support a similar function as starting a VM on boot, or should I delete the App Store app and reinstall from the repository to an earlier version?

**Comment by osy** [comment]
Can you make your script launch UTM, wait a few seconds, then try to use utmctl?

**Reply by EagleDTW** [reply]
> Can you make your script launch UTM, wait a few seconds, then try to use utmctl?

Neither making the script wait for 50 seconds or longer, nor having the shortcut start a different shortcut that then does the utmctl command work...in all cases the VM not found message is always displayed and shortcut errors out.

**Reply by osy** [reply]
Can you try using & to background your script? I wonder if macOS is doing something weird like it won't launch gui apps until the launch scripts are done?

**Reply by EagleDTW** [reply]
> Can you try using & to background your script? I wonder if macOS is doing something weird like it won't launch gui apps until the launch scripts are done?

I'll try to set to background script...the app does open - but possibly the interface to the cli is not actually able to run?

**Reply by EagleDTW** [reply]
> > Can you try using & to background your script? I wonder if macOS is doing something weird like it won't launch gui apps until the launch scripts are done?
> 
> I'll try to set to background script...the app does open - but possibly the interface to the cli is not actually able to run?

@osy Background running this command works while the system is up, but fails silently at boot and doesn't start the servers:
bash -c "nohup sh -c 'utmctl start Servers' &"

**Reply by EagleDTW** [reply, marked as answer]
Apparently this is the answer and confirmed to be working (from stack overflow):

    Start Automator.app;
    Select "Application";
    Click "Show library" in the toolbar (if hidden);
    Add "Run shell script" (from the Actions/Utilities);
    Copy-and-paste your script into the window;
    Test it;
    Save it somewhere: a file called your_name.app will be created);
    Depending your MacOSX version:
        Old versions: Go to System Preferences → Accounts → Login items, or
        Before Ventura version: Go to System Preferences → Users and Groups → Login items (top right), or
        Ventura and after version: Go to System Preferences → General → Login items;
    Add this newly-created app;

Log off, log back in, accept and allow, and you should be done. ;)
https://superuser.com/questions/229773/run-command-on-startup-login-mac-os-x
